pkg/kubelet/userns_manager.go

```diff
@@ -0,0 +1,476 @@
+		return err
+	}
+	if string(content) == "" {
+		return nil
+	}
+
+	_, err = m.parseUserNsFile(pod, content)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// getUserNamespaceMappingsFile returns the path to the file that contains the user namespace configuration.
+func (m *usernsManager) getUserNamespaceMappingsFile(pod types.UID) string {
+	if m.kl == nil {
+		return ""
+	}
+	return path.Join(m.kl.getPodDir(pod), "userns")
+}
+
+// IsSet checks if the specified index is already set.
+func (m *usernsManager) isSet(v uint32) bool {
+	index := v / userNsLength
+	return m.used.isSet(index)
+}
+
+// allocateOne finds a free user namespace and allocate it to the specified pod.
+// The first return value is the first ID in the user namespace, the second returns
+// the length for the user namespace range.
+func (m *usernsManager) allocateOne(pod string) (uint32, uint32, error) {
+	firstZero, err := m.used.findAndSetFirstZero()
+	if err != nil {
+		return 0, 0, fmt.Errorf("could not allocate user namespace: %v", err)
+	}
+	firstID := firstZero * userNsLength
+	m.usedBy[pod] = firstID
+	return firstID, userNsLength, nil
+}
+
+// record stores the user namespace [from; from+length] to the specified pod.
+func (m *usernsManager) record(pod types.UID, from, length uint32) error {
+	if length != userNsLength {
+		return fmt.Errorf("wrong user namespace length %v", length)
+	}
+	if from%userNsLength != 0 {
+		return fmt.Errorf("wrong user namespace offset specified %v", from)
+	}
+	if prevFrom, found := m.usedBy[string(pod)]; found && prevFrom != from {
+		return fmt.Errorf("user namespace range already used by pod %q", pod)
+	}
+
+	// "from" is a ID (UID/GID), set the corresponding userns of size
+	// userNsLength in the bit-array.
+	m.used.set(from / userNsLength)
+	m.usedBy[string(pod)] = from
+	return nil
+}
+
+// Release releases the user namespace allocated to the specified pod.
+func (m *usernsManager) release(pod string) {
+	v, ok := m.usedBy[pod]
+	if !ok {
+		return
+	}
+	delete(m.usedBy, pod)
+
+	m.currPods--
+	m.removed++
+	if m.removed%1000 == 0 {
+		n := make(map[string]uint32)
+		for k, v := range m.usedBy {
+			n[k] = v
+		}
+		m.usedBy = n
+		m.removed = 0
+	}
+	m.used.clear(v / userNsLength)
+}
+
+func (m *usernsManager) parseUserNsFile(pod types.UID, content []byte) (userNs userNamespace, err error) {
+	if err = json.Unmarshal([]byte(content), &userNs); err != nil {
+		err = fmt.Errorf("can't parse file: %w", err)
+		return
+	}
+
+	if len(userNs.UIDMappings) != 1 {
+		err = fmt.Errorf("invalid user namespace configuration: no more than one mapping allowed.")
+		return
+	}
+
+	if len(userNs.UIDMappings) != len(userNs.GIDMappings) {
+		err = fmt.Errorf("invalid user namespace configuration: GID and UID mappings should be identical.")
+		return
+	}
+
+	if userNs.UIDMappings[0] != userNs.GIDMappings[0] {
+		err = fmt.Errorf("invalid user namespace configuration: GID and UID mapping should be identical")
+		return
+	}
```

Comment:
Are these tests enough?

Reply:
good eye, we need to validate each mapping not just the first.  I'll fix it

Reply:
@giuseppe no, please, don't replace it with a loop. We are asserting that there is one element here, why can't we just access it? Also, it will make the checks below more tricky, for no reason IMHO.

@sftim yes, these tests are enough. We assert that the number of entries is 1 for UID/GID and that for that entry, the configs for UIDs and GIDs are identical and they map the root user too. Yes, that covers it all.  Can you please elaborate?

Reply:
To make it clear, how about adding a comment before `if len(userNs.UIDMappings) != 1 {` to explain that if removing this restriction, the following tests will need to change.

I didn't spot the link immediately; a comment will explain the intent there.

Reply:
@sftim not sure I see the point for such a comment.

If that if is removed and nothing else is changed, then the rest is ignored so it has the same behavior as not making any change (just not errored out). I don't think that is a mistake that will happen (the new func doesn't work at all!) nor that new tests won't cover it.

Are you really saying we should add a comment saying: if you remove this if and don't change anything else, then your new functionality won't work, be careful when you add new functionality to make sure it works?

I think I misunderstanding you. What am I missing?